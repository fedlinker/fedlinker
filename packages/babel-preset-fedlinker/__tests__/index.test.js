const babel = require('@babel/core');
const preset = require('../index');

beforeEach(() => {
  process.env = Object.assign(process.env, { BABEL_ENV: 'development' });
});
afterEach(() => {
  process.env = Object.assign(process.env, { BABEL_ENV: 'development' });
});

describe('babel-preset-fedlinker', () => {
  describe('environment', () => {
    const input = `
      import React, { Component } from 'react'
      import propTypes from 'prop-types'
      class Button extends Component {
        render() {
          const { disabled, children } = this.props
          return (
            <button disabled={disabled}>{children}</button>
          )
        }
      }
      Button.propTypes = { disabled: propTypes.bool, children: propTypes.node }
      export default Button
    `;
    const options = {
      babelrc: false,
      presets: [[preset]],
      filename: '/index.jsx',
    };

    test('in development', () => {
      process.env.BABEL_ENV = 'development';
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });

    test('in production', () => {
      process.env.BABEL_ENV = 'production';
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });

    test('in test', () => {
      process.env.BABEL_ENV = 'test';
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });
  });

  describe('static type checker', () => {
    test(`can't enable flow and typescript both`, () => {
      expect(() => {
        babel.transformSync('const a = "a"', {
          babelrc: false,
          presets: [[preset, { flow: true, typescript: true }]],
        });
      }).toThrow();
    });

    test('should throw error if write flow or ts code without enabling one of them', () => {
      const flowInput = `
        // @flow
        function square(n: number): number {
          return n * n;
        }
      `;
      const tsInput = `
        function identity<T>(arg: T): T {
          return arg;
        }
      `;
      const options = {
        babelrc: false,
        presets: [[preset]],
      };
      expect(() => {
        babel.transformSync(flowInput, options);
      }).toThrow();
      expect(() => {
        babel.transformSync(tsInput, options);
      }).toThrow();
    });

    test('transform flow syntax correctly', () => {
      const input = `
        // @flow
        function square(n: number): number {
          return n * n;
        }
      `;
      const options = {
        babelrc: false,
        presets: [[preset, { flow: true }]],
        filename: '/index.js',
      };
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });

    test('transform typescript syntax correctly', () => {
      const input = `
        function identity<T>(arg: T): T {
          return arg;
        }
      `;
      const options = {
        babelrc: false,
        presets: [[preset, { typescript: true }]],
        filename: '/index.ts',
      };
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });
  });

  describe('proposals', () => {
    test('should throw error when disable proposals', () => {
      const input = `
        class Bork {
          static a = 'foo';
        }
      `;
      const options = {
        babelrc: false,
        presets: [[preset, { proposals: false }]],
      };
      expect(() => {
        babel.transformSync(input, options);
      }).toThrow();
    });

    test('minimal proposals', () => {
      const input = `
        @decorator
        class Class { static prop = 'prop'; }
      `;
      const options = {
        babelrc: false,
        presets: [[preset, { proposals: 'minimal' }]],
        filename: '/index.js',
      };
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });

    test('all proposals', () => {
      const input = `
        async function asyncFunction() { await awaitFunciton(); }

        @decorator
        class Class { static prop = 'prop'; }

        const doExpression = do { 'value'; };

        const nullishCoalescing = object.foo ?? 'bar';

        boundObject::functionBind

        function* functionSent() { console.log(function.sent); }

        logicalAssignment ||= otherVariable;

        const numericSeparator = 1_000_000_000_000;

        const optionalChaining = a?.b?.c?.d;

        partialApplication(?, param);

        const pipelineResult = "value" |> processFunction

        class PrivateMethodClass { #privateProp = 'value'; }

        const throwExpression = throw new Error("message")

        export exportDefaultFrom from 'module-name';

        export * as exportNamespaceFrom from 'module-name';
      `;
      const options = {
        babelrc: false,
        presets: [[preset, { proposals: 'all' }]],
        filename: '/index.js',
      };
      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });
  });

  describe('polyfills', () => {
    test('should inject default polyfills and user polyfills in entry file', () => {
      const options = {
        babelrc: false,
        presets: [[preset, { entry: '/entry.js', polyfills: ['module-name'] }]],
        filename: '/entry.js',
      };
      const { code } = babel.transformSync('', options);
      expect(code).toMatchSnapshot();
    });

    test('should not inject default polyfills when disable shim', () => {
      const options = {
        babelrc: false,
        presets: [[preset, { entry: '/entry.js', shim: false }]],
        filename: '/entry.js',
      };
      const { code } = babel.transformSync('', options);
      expect(code).not.toMatch('import');
      expect(code).not.toMatch('require');
    });
  });
});
