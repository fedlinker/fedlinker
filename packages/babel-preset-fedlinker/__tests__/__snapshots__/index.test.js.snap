// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-preset-fedlinker environment in development 1`] = `
"import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";
import _createClass from \\"@babel/runtime-corejs3/helpers/esm/createClass\\";
import _possibleConstructorReturn from \\"@babel/runtime-corejs3/helpers/esm/possibleConstructorReturn\\";
import _getPrototypeOf from \\"@babel/runtime-corejs3/helpers/esm/getPrototypeOf\\";
import _inherits from \\"@babel/runtime-corejs3/helpers/esm/inherits\\";
var _jsxFileName = \\"/index.jsx\\";
import React, { Component } from 'react';
import propTypes from 'prop-types';

var Button =
/*#__PURE__*/
function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    _classCallCheck(this, Button);

    return _possibleConstructorReturn(this, _getPrototypeOf(Button).apply(this, arguments));
  }

  _createClass(Button, [{
    key: \\"render\\",
    value: function render() {
      var _this$props = this.props,
          disabled = _this$props.disabled,
          children = _this$props.children;
      return React.createElement(\\"button\\", {
        disabled: disabled,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 8
        },
        __self: this
      }, children);
    }
  }]);

  return Button;
}(Component);

Button.propTypes = {
  disabled: propTypes.bool,
  children: propTypes.node
};
export default Button;"
`;

exports[`babel-preset-fedlinker environment in production 1`] = `
"import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";
import _createClass from \\"@babel/runtime-corejs3/helpers/esm/createClass\\";
import _possibleConstructorReturn from \\"@babel/runtime-corejs3/helpers/esm/possibleConstructorReturn\\";
import _getPrototypeOf from \\"@babel/runtime-corejs3/helpers/esm/getPrototypeOf\\";
import _inherits from \\"@babel/runtime-corejs3/helpers/esm/inherits\\";
import React, { Component } from 'react';

var Button =
/*#__PURE__*/
function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    _classCallCheck(this, Button);

    return _possibleConstructorReturn(this, _getPrototypeOf(Button).apply(this, arguments));
  }

  _createClass(Button, [{
    key: \\"render\\",
    value: function render() {
      var _this$props = this.props,
          disabled = _this$props.disabled,
          children = _this$props.children;
      return React.createElement(\\"button\\", {
        disabled: disabled
      }, children);
    }
  }]);

  return Button;
}(Component);

export default Button;"
`;

exports[`babel-preset-fedlinker environment in test 1`] = `
"\\"use strict\\";

var _interopRequireWildcard = require(\\"@babel/runtime-corejs3/helpers/interopRequireWildcard\\");

var _interopRequireDefault = require(\\"@babel/runtime-corejs3/helpers/interopRequireDefault\\");

var _Object$defineProperty = require(\\"@babel/runtime-corejs3/core-js-stable/object/define-property\\");

_Object$defineProperty(exports, \\"__esModule\\", {
  value: true
});

exports.default = void 0;

var _react = _interopRequireWildcard(require(\\"react\\"));

var _propTypes = _interopRequireDefault(require(\\"prop-types\\"));

var _jsxFileName = \\"/index.jsx\\";

class Button extends _react.Component {
  render() {
    const {
      disabled,
      children
    } = this.props;
    return _react.default.createElement(\\"button\\", {
      disabled: disabled,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 8
      },
      __self: this
    }, children);
  }

}

Button.propTypes = {
  disabled: _propTypes.default.bool,
  children: _propTypes.default.node
};
var _default = Button;
exports.default = _default;"
`;

exports[`babel-preset-fedlinker polyfills should inject default polyfills and user polyfills in entry file 1`] = `
"import \\"core-js/features/array/iterator\\";
import \\"core-js/features/promise\\";
import \\"core-js/features/promise/finally\\";
import \\"core-js/features/object/assign\\";
import \\"core-js/features/array/from\\";
import \\"core-js/features/symbol\\";
import \\"core-js/features/set\\";
import \\"core-js/features/map\\";
import \\"raf/polyfill\\";
import \\"whatwg-fetch\\";
import \\"abortcontroller-polyfill/dist/polyfill-patch-fetch\\";
import \\"module-name\\";"
`;

exports[`babel-preset-fedlinker proposals all proposals 1`] = `
"import _regeneratorRuntime from \\"@babel/runtime-corejs3/regenerator\\";
import _Object$defineProperty from \\"@babel/runtime-corejs3/core-js-stable/object/define-property\\";
import _classPrivateFieldLooseKey from \\"@babel/runtime-corejs3/helpers/esm/classPrivateFieldLooseKey\\";
import _skipFirstGeneratorNext from \\"@babel/runtime-corejs3/helpers/esm/skipFirstGeneratorNext\\";
import \\"regenerator-runtime/runtime\\";
import _bindInstanceProperty from \\"@babel/runtime-corejs3/core-js-stable/instance/bind\\";
import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";
import _asyncToGenerator from \\"@babel/runtime-corejs3/helpers/esm/asyncToGenerator\\";

var _class, _class2, _temp, _object$foo, _context, _context2, _a, _a$b, _a$b$c, _param, _partialApplication, _value;

function asyncFunction() {
  return _asyncFunction.apply(this, arguments);
}

function _asyncFunction() {
  _asyncFunction = _asyncToGenerator(
  /*#__PURE__*/
  _regeneratorRuntime.mark(function _callee() {
    return _regeneratorRuntime.wrap(function _callee$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return awaitFunciton();

          case 2:
          case \\"end\\":
            return _context3.stop();
        }
      }
    }, _callee);
  }));
  return _asyncFunction.apply(this, arguments);
}

var Class = decorator(_class = (_temp = _class2 = function Class() {
  _classCallCheck(this, Class);
}, _class2.prop = 'prop', _temp)) || _class;

var doExpression = 'value';
var nullishCoalescing = (_object$foo = object.foo) !== null && _object$foo !== void 0 ? _object$foo : 'bar';

_bindInstanceProperty(_context2 = (_context = boundObject, functionBind)).call(_context2, _context);

function functionSent() {
  return _functionSent2.apply(this, arguments);
}

function _functionSent2() {
  _functionSent2 = _skipFirstGeneratorNext(
  /*#__PURE__*/
  _regeneratorRuntime.mark(function _callee2() {
    var _functionSent;

    return _regeneratorRuntime.wrap(function _callee2$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return;

          case 2:
            _functionSent = _context4.sent;
            console.log(_functionSent);

          case 4:
          case \\"end\\":
            return _context4.stop();
        }
      }
    }, _callee2);
  }));
  return _functionSent2.apply(this, arguments);
}

logicalAssignment || (logicalAssignment = otherVariable);
var numericSeparator = 1000000000000;
var optionalChaining = (_a = a) === null || _a === void 0 ? void 0 : (_a$b = _a.b) === null || _a$b === void 0 ? void 0 : (_a$b$c = _a$b.c) === null || _a$b$c === void 0 ? void 0 : _a$b$c.d;
_partialApplication = partialApplication, _param = param, function partialApplication(_argPlaceholder) {
  return _partialApplication(_argPlaceholder, _param);
};
var pipelineResult = (_value = \\"value\\", processFunction(_value));

var PrivateMethodClass = function PrivateMethodClass() {
  _classCallCheck(this, PrivateMethodClass);

  _Object$defineProperty(this, _privateProp, {
    writable: true,
    value: 'value'
  });
};

var _privateProp = _classPrivateFieldLooseKey(\\"privateProp\\");

var throwExpression = function (e) {
  throw e;
}(new Error(\\"message\\"));

import _exportDefaultFrom from 'module-name';
export { _exportDefaultFrom as exportDefaultFrom };
import * as _exportNamespaceFrom from 'module-name';
export { _exportNamespaceFrom as exportNamespaceFrom };"
`;

exports[`babel-preset-fedlinker proposals minimal proposals 1`] = `
"import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";

var _class, _class2, _temp;

var Class = decorator(_class = (_temp = _class2 = function Class() {
  _classCallCheck(this, Class);
}, _class2.prop = 'prop', _temp)) || _class;"
`;

exports[`babel-preset-fedlinker static type checker transform flow syntax correctly 1`] = `
"function square(n) {
  return n * n;
}"
`;

exports[`babel-preset-fedlinker static type checker transform typescript syntax correctly 1`] = `
"function identity(arg) {
  return arg;
}"
`;
